src/StreamJsonRpc/Reflection/RpcTargetInfo.cs

```diff
@@ -0,0 +1,685 @@
+        /// </summary>
+        private List<EventReceiver>? eventReceivers;
+
+        /// <summary>
+        /// A lazily-initialized list of objects to dispose of when the JSON-RPC connection drops.
+        /// </summary>
+        private List<object>? localTargetObjectsToDispose;
+
+        internal RpcTargetInfo(JsonRpc jsonRpc)
+        {
+            this.jsonRpc = jsonRpc;
+        }
+
+        private TraceSource TraceSource => this.jsonRpc.TraceSource;
+
+        private object SyncObject => this.targetRequestMethodToClrMethodMap;
+
+        public async ValueTask DisposeAsync()
+        {
+            if (this.localTargetObjectsToDispose is object)
+            {
+                List<Exception>? exceptions = null;
+                foreach (object target in this.localTargetObjectsToDispose)
+                {
+                    // We're calling Dispose on the target objects, so switch to the user-supplied SyncContext for those target objects.
+                    await this.jsonRpc.SynchronizationContextOrDefault;
+
+                    try
+                    {
+                        // Arrange to dispose of the target when the connection is closed.
+                        if (target is System.IAsyncDisposable asyncDisposableTarget)
+                        {
+                            await asyncDisposableTarget.DisposeAsync().ConfigureAwait(false);
+                        }
+                        else if (target is Microsoft.VisualStudio.Threading.IAsyncDisposable vsAsyncDisposableTarget)
+                        {
+                            await vsAsyncDisposableTarget.DisposeAsync().ConfigureAwait(false);
+                        }
+                        else if (target is IDisposable disposableTarget)
+                        {
+                            disposableTarget.Dispose();
+                        }
+                    }
+#pragma warning disable CA1031 // Do not catch general exception types
+                    catch (Exception ex)
+#pragma warning restore CA1031 // Do not catch general exception types
+                    {
+                        if (exceptions is null)
+                        {
+                            exceptions = new List<Exception>();
+                        }
+
+                        exceptions.Add(ex);
+                    }
+                }
+
+                if (exceptions is object)
+                {
+                    if (exceptions.Count == 1)
+                    {
+                        ExceptionDispatchInfo.Capture(exceptions[0]).Throw();
+                    }
+
+                    throw new AggregateException(exceptions);
+                }
+            }
+        }
+
+        internal static MethodNameMap GetMethodNameMap(TypeInfo type)
+        {
+            MethodNameMap? map;
+            lock (MethodNameMaps)
```

Comment:
I assume this means gains by reduced GC pressure outweighs the overhead of the lock?

Reply:
Oh yes, very much so. 18% faster, remember. 😏
And ya, the reflection that we skip here is expensive in terms of GC pressure. Almost every API we call creates _arrays_, even if we just wanted one object. Even `GetCustomAttribute` calls another API that returns an array before returning a single element from it.

Reply:
Also note, these are all optimistic locking. We take it just long enough to peek, then we release the lock to do the work. As a result, this lock is not held for very long at all => minimal lock contention.